erts/emulator/beam/beam_file.c

```diff
@@ -607,7 +607,11 @@ static int parse_type_chunk(BeamFile *beam, IFF_Chunk *chunk) {
     beamreader_init(chunk->data, chunk->size, &reader);
 
     LoadAssert(beamreader_read_i32(&reader, &version));
-    LoadAssert(version == BEAM_TYPES_VERSION);
+    if (version != BEAM_TYPES_VERSION) {
```

Comment:
Probably not by default. It could get very annoying for systems that intentionally mix BEAM files compiled by different versions of the compiler. But there should be some way or some tool so that one could get aware that type-based optimizations are no longer active.